hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java

```diff
@@ -2613,6 +2615,73 @@ public ServiceInfoEx getServiceInfo() throws IOException {
     return new ServiceInfoEx(getServiceList(), caCertPem);
   }
 
+  private final List<String> finalizationMsgs = new ArrayList<>();
+  private UpgradeFinalizationStatus.Status finalizationStatus =
+      UpgradeFinalizationStatus.Status.FINALIZATION_REQUIRED;
+
+  @Override
+  public UpgradeFinalizationStatus finalizeUpgrade(String upgradeClientID)
+      throws IOException {
+    if (!finalizationStatus
+        .equals(UpgradeFinalizationStatus.Status.FINALIZATION_REQUIRED)){
+      throw new OMException("Finalization is not needed.", INVALID_REQUEST);
+    }
+    finalizationStatus = UpgradeFinalizationStatus.Status.STARTING_FINALIZATION;
+    UpgradeFinalizationStatus status = UpgradeFinalizationStatus.newBuilder()
+        .setStatus(finalizationStatus)
+        .build();
+    LOG.info("finalizeUpgrade initiated by client: {}.", upgradeClientID);
+    if (isLeader()) {
+      finalizationMsgs.add("Finalization started.");
+      finalizationStatus =
+          UpgradeFinalizationStatus.Status.FINALIZATION_IN_PROGRESS;
+
+      new Thread(() -> {
+        LOG.info("Finalization thread started.");
+        int i = 0;
+        Random random = new Random(0xafaf);
+        while (i < 50) {
+          int rand = random.nextInt(Math.min(10, 50 - i)) + 1;
+          synchronized (finalizationMsgs) {
+            LOG.info("Emitting {} messages", rand);
+            for (int j = 0; j < rand; j++) {
+              LOG.info("Upgrade MSG: {} - added.", "Message " + i + ".");
+              finalizationMsgs.add("Message " + i + ".");
+              i++;
+            }
+          }
+          try {
+            int sleep = random.nextInt(1200);
+            LOG.info("Sleeping {}ms before emit messages again.", sleep);
+            Thread.sleep(sleep);
+          } catch (InterruptedException e) {
+            LOG.info("Finalization thread interrupted.", e);
+            return;
+          }
+        }
+        LOG.info("Finalization done.");
+        finalizationStatus = UpgradeFinalizationStatus.Status.FINALIZATION_DONE;
+      }, "Finalization-Thread").start();
+    }
+    return status;
+  }
+
+  @Override
+  public UpgradeFinalizationStatus queryUpgradeFinalizationProgress(
+      String upgradeClientID, boolean force
```

Comment:
force parameter is not actually used here, this parameter is related to the client takeover behavior.

Reply:
In the PR that is forming for HDDS-4172 (the server side implementation), the parameter should be used to validate if we are getting the progress monitoring request from the same client, and error out if we do not.
But in the demo code, I haven't spent time on this part.
On the other hand as from the client side this has changed to be the -takeover parameter, I am renaming the parameter name to takeover here and in proto as well.

Reply:
I have seen that force has been renamed to -takeover. I will look into server side implementation in HDDS-4172 once you attach the new PR there.